crates/core/tedge_api/src/entity.rs

```diff
@@ -0,0 +1,90 @@
+use crate::entity_store::EntityType;
+use crate::mqtt_topics::EntityTopicId;
+use crate::mqtt_topics::TopicIdError;
+use serde::Deserialize;
+use serde::Serialize;
+
+/// Represents externally provided unique ID of an entity.
+///
+/// Although this struct doesn't enforce any restrictions for the values,
+/// the consumers may impose restrictions on the accepted values.
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct EntityExternalId(String);
+
+impl AsRef<str> for EntityExternalId {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+// XXX: As `EntityExternalId` is used as a part of cloudbound MQTT topic, it
+// can't contain characters invalid in topics, i.e. `+` and `#`. ([MQTT-4.7]).
+// If it's derived from a MQTT topic, this holds, but if created from a string,
+// this isn't checked, which is invalid!
+impl From<&str> for EntityExternalId {
+    fn from(val: &str) -> Self {
+        Self(val.to_string())
+    }
+}
+
+impl From<&String> for EntityExternalId {
+    fn from(val: &String) -> Self {
+        Self(val.to_string())
+    }
+}
+
+impl From<String> for EntityExternalId {
+    fn from(val: String) -> Self {
+        Self(val)
+    }
+}
+
+impl From<EntityExternalId> for String {
+    fn from(value: EntityExternalId) -> Self {
+        value.0
+    }
+}
+
+impl From<&EntityExternalId> for String {
+    fn from(value: &EntityExternalId) -> Self {
+        value.0.clone()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EntityMetadata {
```

Comment:
As mentioned in https://github.com/thin-edge/thin-edge.io/pull/3230#discussion_r1878052707, the entity store and entity cache needs to maintain their own versions of entity metadata as the latter would be much lighter with minimal metadata maintained at the mapper level. Ideally the `EntityMetadata` maintained by the mapper should have been as minimal as the XID. But, I was forced to store a bit more info like the entity type, display name and display type just because of the way we map service health status messages by publishing the service creation message itself(which requires all that metadata) for status updates as well. One way to eliminate caching this additional metadata at the mapper level is by making it query the entity store contents from the agent, whenever it requires all that metadata. But that extra HTTP call during the mapping complicates the conversion logic and I just chose caching extra metadata instead of that added complexity. But, open to revisit that decision in a follow-up PR dedicated for that change as the unit test impact would be quite big.

Reply:
> But that extra HTTP call during the mapping complicates the conversion logic and I just chose caching extra metadata instead of that added complexity. 

Agree, caching registration messages is simpler than late-requests to the entity store.

But could we share the type of cached data? Say raw registration messages? Even if some fields are not currently used in the mapper.

Reply:
> But could we share the type of cached data? Say raw registration messages? Even if some fields are not currently used in the mapper.

Yeah, that is an option. I introduced a different struct to clearly represent the minimal set of things that a mapper would need to do the mapping. And also, there were minor things like the external id (`@id`) being an `Option` in the actual `EntityRegistrationMessage`, but the same being a mandatory field in the entity cache. But yeah, if we want to avoid having too many "metadata representations" across the crates, I can reuse some of the existing ones with implicit assumptions about fields like `@id`.

Reply:
We really have to reduce the number of similar but different structs.